torchaudio_contrib/functional.py

```diff
@@ -1,106 +1,133 @@
 import torch
-import numpy as np
+import math
+import torch.nn.functional as F
 
 
-def _get_time_values(sig_length, sr, hop):
+def stft_defaults(n_fft, hop_length, len_win, window):
     """
-    Get the time axis values given the signal length, sample
-    rate and hop size.
+    Should function outisde torchaudio_contrib.stft since
+    torchaudio_contrib.STFT will use it outisde of forward() (?).
     """
-    return torch.linspace(0, sig_length/sr, sig_length//hop+1)
+    hop_length = n_fft // 4 if hop_length is None else hop_length
 
+    if window is None:
+        length = n_fft if len_win is None else len_win
+        window = torch.hann_window(length)
+        if not isinstance(window, torch.Tensor):
+            raise TypeError('window must be a of type torch.Tensor')
 
-def _get_freq_values(n_fft, sr):
-    """
-    Get the frequency axis values given the number of FFT bins
-    and sample rate.
-    """
-    return torch.linspace(0, sr/2, n_fft//2 + 1)
-
+    return n_fft, hop_length, window
 
-def get_spectrogram_axis(sig_length, sr, n_fft=2048, hop=512):
-    t = _get_time_values(sig_length, sr, hop)
-    f = _get_freq_values(n_fft, sr)
-    return t, f
 
+def _stft(x, n_fft, hop_length, window, pad, pad_mode, **kwargs):
+    """
+    Wrap torch.stft allowing for multi-channel stft.
 
-def STFT(sig, n_fft=2048, hop=None, window=None, **kwargs):
-    """A wrapper for torch.stft with some preset parameters.
-    Returned value keeps both real and imageinary parts.
-    For STFT magnitude, see spectrogram
+    Args:
+        x (Tensor): Tensor of audio of size (channel, signal)
+            or (batch, channel, signal).
+        n_fft (int): FFT window size.
+        hop_length (int): Number audio of frames between STFT columns.
+        len_win (int): Size of stft window.
+        window (Tensor): 1-D tensor.
+        pad (int): Amount of padding to apply to signal.
+        pad_mode: padding method (see torch.nn.functional.pad).
+        **kwargs: Other torch.stft parameters, see torch.stft for more details.
 
-    input -> (batch, channel, time) or (channel, time)
-    output -> (batch, channel, freq, hop, complex) or (channel, freq, hop, complex)
+    Returns:
+        Tensor: (batch, channel, freq, hop, complex)
+            or (channel, freq, hop, complex)
 
     """
-    if hop is None:
-        hop = n_fft // 4
-    if window is None:
-        window = torch.hann_window(n_fft)
-
-    if sig.dim() == 3:
-        batch, channel, time = sig.size()
-        out_shape = [batch, channel, n_fft//2+1, -1, 2]
-    elif sig.dim() == 2:
-        channel, time = sig.size()
-        out_shape = [channel, n_fft//2+1, -1, 2]
+
+    # (!) Only 3D, 4D, 5D padding with non-constant
+    # padding are supported for now.
+    if pad > 0:
+        x = F.pad(x, (pad, pad), pad_mode)
```

Comment:
We shouldn't pad here, we should defer this to `torch.stft`. It may have a faster implementation sometime.

Reply:
Yeah I was following [torchaudio](https://github.com/pytorch/audio/blob/ea7fcf080020f27da2f7675f01f94b84bae4c922/torchaudio/transforms.py#L201) where they allow for explicit constant padding.

Reply:
But you can do constant padding with `torch.stft` as well, no?